/* Copyright 2019 The Waymo Open Dataset Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

syntax = "proto2";

package waymo.open_dataset;

import "waymo_open_dataset/protos/keypoint.proto";

message Label {
  // Upright box, zero pitch and roll.
  message Box {
    // Box coordinates in vehicle frame.
    optional double center_x = 1;
    optional double center_y = 2;
    optional double center_z = 3;

    // Dimensions of the box. length: dim x. width: dim y. height: dim z.
    optional double length = 5;
    optional double width = 4;
    optional double height = 6;

    // The heading of the bounding box (in radians).  The heading is the angle
    // required to rotate +x to the surface normal of the box front face. It is
    // normalized to [-pi, pi).
    optional double heading = 7;

    enum Type {
      TYPE_UNKNOWN = 0;
      // 7-DOF 3D (a.k.a upright 3D box).
      TYPE_3D = 1;
      // 5-DOF 2D. Mostly used for laser top down representation.
      TYPE_2D = 2;
      // Axis aligned 2D. Mostly used for image.
      TYPE_AA_2D = 3;
    }
  }

  optional Box box = 1;

  message Metadata {
    optional double speed_x = 1;
    optional double speed_y = 2;
    optional double accel_x = 3;
    optional double accel_y = 4;
  }
  optional Metadata metadata = 2;

  enum Type {
    TYPE_UNKNOWN = 0;
    TYPE_VEHICLE = 1;
    TYPE_PEDESTRIAN = 2;
    TYPE_SIGN = 3;
    TYPE_CYCLIST = 4;
  }
  optional Type type = 3;
  // Object ID.
  optional string id = 4;

  // The difficulty level of this label. The higher the level, the harder it is.
  enum DifficultyLevel {
    UNKNOWN = 0;
    LEVEL_1 = 1;
    LEVEL_2 = 2;
  }

  // Difficulty level for detection problem.
  optional DifficultyLevel detection_difficulty_level = 5;
  // Difficulty level for tracking problem.
  optional DifficultyLevel tracking_difficulty_level = 6;

  // The total number of lidar points in this box.
  optional int32 num_lidar_points_in_box = 7;

  oneof keypoints_oneof {
    // Used if the Label is a part of `Frame.laser_labels`.
    keypoints.LaserKeypoints laser_keypoints = 8;
    // Used if the Label is a part of `Frame.camera_labels`.
    keypoints.CameraKeypoints camera_keypoints = 9;
  }

  // Information to cross reference between labels for different modalities.
  message Association {
    // Currently only CameraLabels store information about associated lidar
    // objects.
    optional string laser_object_id = 1;
  }
  optional Association association = 10;

}

// Non-self-intersecting 2d polygons. This polygon is not necessarily convex.
message Polygon2dProto {
  repeated double x = 1;
  repeated double y = 2;

  // A globally unique ID.
  optional string id = 3;
}

// This is not finalized.
// TODO(yinzhou,rqi): Update this once  the mapping is finalized.
message Segmentation {
  enum Type {
    TYPE_NOT_LABELED = 0;
    TYPE_MOVABLE_OBJECT = 1;
    TYPE_VEHICLE = 2;
    TYPE_SMALL_VEHICLE = 3;
    TYPE_CAR = 4;
    TYPE_THREE_WHEEL_VEHICLE = 5;
    TYPE_OTHER_SMALL_VEHICLE = 47;
    TYPE_LARGE_VEHICLE = 6;
    TYPE_FREIGHT_TRUCK = 7;
    TYPE_BUS = 8;
    TYPE_OTHER_LARGE_VEHICLE = 48;
    TYPE_TWO_WHEEL_VEHICLE = 9;
    TYPE_MOTORCYCLIST = 10;
    TYPE_CYCLIST = 11;
    TYPE_OTHER_TWO_WHEEL_VEHICLE = 49;
    TYPE_TRAILER = 12;
    TYPE_OPEN_DOOR = 63;
    TYPE_VEHICLE_ATTACHMENT = 64;
    TYPE_PROTRUSION = 65;
    TYPE_OTHER_VEHICLE = 50;
    TYPE_PEDESTRIAN = 13;
    TYPE_ADULT = 14;
    TYPE_CHILD = 15;
    TYPE_OTHER_PEDESTRIAN = 51;
    TYPE_PEDESTRIAN_OBJECT = 16;
    TYPE_HOLDING_SIGN = 17;
    TYPE_OTHER_PEDESTRIAN_OBJECT = 52;
    TYPE_ANIMAL = 18;
    TYPE_BIRD = 71;
    TYPE_CAT = 72;
    TYPE_COW = 73;
    TYPE_DEER = 74;
    TYPE_DOG = 75;
    TYPE_HORSE = 76;
    TYPE_OTHER_ANIMAL = 77;
    TYPE_OTHER_MOVABLE_OBJECT = 53;
    TYPE_IMPERMANENT_OBJECT = 19;
    TYPE_CONSTRUCTION_CONE_POLE = 20;
    TYPE_TEMPORARY_SIGN = 21;
    TYPE_TEMPORARY_POLE = 22;
    TYPE_BICYCLE = 23;
    TYPE_MOTORCYCLE = 24;
    TYPE_AUTOMOTIVE_DEBRIS = 78;
    TYPE_SHOPPING_CART = 79;
    TYPE_OTHER_IMPERMANENT_OBJECT = 54;
    TYPE_PERMANENT_OBJECT = 25;
    TYPE_BUILDING = 26;
    TYPE_GARAGE_DOOR = 62;
    TYPE_OTHER_BUILDING = 66;
    TYPE_SIGN = 27;
    TYPE_POLE = 28;
    TYPE_TRAFFICLIGHT = 29;
    TYPE_VEGETATION = 30;
    TYPE_TREE_TRUNK = 31;
    TYPE_OTHER_VEGETATION = 55;
    TYPE_CURB = 32;
    TYPE_CURB_CUT = 61;
    TYPE_OTHER_CURB = 69;
    TYPE_JERSEY_BARRIER = 70;
    TYPE_OTHER_PERMANENT_OBJECT = 56;
    TYPE_SDC = 33;
    TYPE_SKY = 34;

    // This is a type used for far away places where it's almost invisible due
    // to Night/Snowy/Rainy/Foggy. This type should not be used for
    // blurry/uncertain object boundaries (where we would use best guess
    // strategy).
    TYPE_UNKNOWN = 35;

    TYPE_DRIVABLE_SURFACE = 36;
    TYPE_ROAD = 37;
    TYPE_LANE_MARKER = 38;
    TYPE_ROAD_MARKER = 39;
    TYPE_BUMP = 40;
    TYPE_CATEYE = 41;
    TYPE_OTHER_ROAD = 57;
    TYPE_EMBEDDED_RAILTRACK = 42;
    TYPE_OTHER_DRIVABLE_SURFACE = 58;
    TYPE_WALKABLE_SURFACE = 43;
    TYPE_SIDEWALK = 44;
    TYPE_DRIVABLE_SURFACE_OVER_SIDEWALK = 67;
    TYPE_OTHER_SIDEWALK = 68;
    TYPE_OTHER_WALKABLE_SURFACE = 59;

    TYPE_SPURIOUS = 45;
    TYPE_REFLECTION = 46;
    TYPE_OTHER_SPURIOUS = 60;
  }
}
